meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/LockClient.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.helix.metaclient.datamodel.DataRecord;
+import org.apache.helix.metaclient.exception.MetaClientException;
+import org.apache.helix.metaclient.factories.MetaClientConfig;
+import org.apache.helix.metaclient.factories.MetaClientFactory;
+import org.apache.helix.metaclient.impl.zk.factory.ZkMetaClientConfig;
+import org.apache.helix.metaclient.impl.zk.factory.ZkMetaClientFactory;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class LockClient implements LockClientInterface, AutoCloseable {
+  private final MetaClientInterface<LockInfo> _metaClient;
+
+  public LockClient(MetaClientConfig config) {
+    if (config.getStoreType().equals(MetaClientConfig.StoreType.ZOOKEEPER)) {
+      ZkMetaClientConfig zkMetaClientConfig = new ZkMetaClientConfig.ZkMetaClientConfigBuilder().
+          setConnectionAddress(config.getConnectionAddress())
+          // Currently only support ZNRecordSerializer. TODO: make this configurable
+          .setZkSerializer((new ZNRecordSerializer()))
+          .build();
+      _metaClient = new ZkMetaClientFactory().getMetaClient(zkMetaClientConfig);
+    } else {
+      throw new MetaClientException("Unsupported store type: " + config.getStoreType());
+    }
+    _metaClient.connect();
+  }
+
+  public LockClient(MetaClientInterface<LockInfo> client) {
+    _metaClient = client;
+    _metaClient.connect();
+  }
+
+  @Override
+  public void acquireLock(String key, LockInfo lockInfo, MetaClientInterface.EntryMode mode) {
+    _metaClient.create(key, lockInfo, mode);
+  }
+
+  @Override
+  public void acquireLockWithTTL(String key, LockInfo lockInfo, long ttl) {
+    _metaClient.createWithTTL(key, lockInfo, ttl);
+  }
+
+  @Override
+  public void renewTTLLock(String key) {
+    _metaClient.renewTTLNode(key);
+  }
+
+  @Override
+  public void releaseLock(String key) {
+    MetaClientInterface.Stat stat = _metaClient.exists(key);
+    if (stat != null) {
+      int version = stat.getVersion();
+      List<Op> ops = Arrays.asList(
+          Op.check(key, version),
+          Op.delete(key, version));
+      _metaClient.transactionOP(ops);
+      if (_metaClient.exists(key) != null) {
+        throw new MetaClientException("Failed to release lock for key: " + key);
+      }
+    }
+  }
+
+  @Override
+  public LockInfo retrieveLock(String key) {
+    MetaClientInterface.Stat stat = _metaClient.exists(key);
+    if (stat == null) {
+      return null;
+    }
+    //Create a new DataRecord from underlying record
+    DataRecord dataRecord = new DataRecord(_metaClient.get(key));
+    //Create a new LockInfo from DataRecord
+    LockInfo lockInfo = new LockInfo(dataRecord);
+    //Synchronize the lockInfo with the stat
+    lockInfo.setGrantedAt(stat.getCreationTime());
+    lockInfo.setLastRenewedAt(stat.getModifiedTime());
+    return lockInfo;
+  }
+
+  @Override
+  public void close() {
+    _metaClient.disconnect();
```

Comment:
If user uses the constructor `public LockClient(MetaClientInterface<LockInfo> client)`, the metaclient shouldn't be closed, it will have external unexpected effect.

Reply:
I'm not quite sure I understand. The metaclient will only be closed when close is called, which is the desired effect